File: src/tui/agent-select.ts
import * as p from "@clack/prompts";
import { AGENT_REGISTRY, type AgentId } from "../agents.js";
import { parseLaunchCommand } from "./parse.js";

const UNSET = "__unset__";
const CUSTOM = "__custom__";

const BUILTIN_OPTIONS = (Object.keys(AGENT_REGISTRY) as AgentId[]).map((id) => ({
  value: id,
  label: AGENT_REGISTRY[id].kind === "none" ? `${id} (no multi-root)` : id,
}));

const SELECT_OPTIONS = [
  { value: UNSET, label: "auto-detect (from launchCommand)" },
  ...BUILTIN_OPTIONS,
  { value: CUSTOM, label: "custom (configure addDirArgs)" },
];

export type AgentChoiceResult =
  | { cancelled: true }
  | {
      cancelled: false;
      agent: AgentId | undefined;
      addDirArgs: string[] | undefined;
    };

export interface PromptAgentChoiceOptions {
  message: string;
  currentAgent?: AgentId;
  currentAddDirArgs?: readonly string[];
}

export async function promptAgentChoice(
  options: PromptAgentChoiceOptions,
): Promise<AgentChoiceResult> {
  const { message, currentAgent, currentAddDirArgs } = options;
  const initialValue = currentAgent ?? (currentAddDirArgs ? CUSTOM : UNSET);

  const choice = await p.select({
    message,
    options: SELECT_OPTIONS,
    initialValue,
  });
  if (p.isCancel(choice)) return { cancelled: true };

  if (choice === UNSET) {
    return { cancelled: false, agent: undefined, addDirArgs: undefined };
  }
  if (choice === CUSTOM) {
    const input = await p.text({
      message: "addDirArgs template (space-separated, use {path} for sibling)",
      placeholder: "--add-dir {path}",
      initialValue: currentAddDirArgs
        ? currentAddDirArgs.join(" ")
        : "--add-dir {path}",
      validate: (v) => (v.includes("{path}") ? undefined : "must contain {path}"),
    });
    if (p.isCancel(input)) return { cancelled: true };
    return {
      cancelled: false,
      agent: undefined,
      addDirArgs: parseLaunchCommand(input as string),
    };
  }
  return {
    cancelled: false,
    agent: choice as AgentId,
    addDirArgs: undefined,
  };
}

File: internal/tui/agentselect.go
package tui

import (
	"errors"
	"strings"

	"github.com/charmbracelet/huh"

	"github.com/OysterD3/wtmux/internal/agents"
)

// AgentChoice is the result of AgentSelect. The TUI surfaces three states:
// "auto-detect" (Agent="" and AddDirArgs=nil), one of the registry agents
// (Agent set, AddDirArgs=nil), or a custom addDirArgs template
// (Agent="" and AddDirArgs populated). Cancelled distinguishes Esc/Ctrl-C
// from a real selection.
type AgentChoice struct {
	Cancelled  bool
	Agent      agents.AgentID
	AddDirArgs []string
}

// AgentSelectInput configures AgentSelect.
type AgentSelectInput struct {
	Title             string
	CurrentAgent      agents.AgentID
	CurrentAddDirArgs []string
}

const (
	agentSelectUnset  = "__unset__"
	agentSelectCustom = "__custom__"
)

// AgentSelect runs the agent picker. Returns Cancelled=true when the user
// aborts the form (Esc/Ctrl-C).
func AgentSelect(in AgentSelectInput) (AgentChoice, error) {
	options := []huh.Option[string]{
		huh.NewOption("auto-detect (from launchCommand)", agentSelectUnset),
	}
	for _, id := range agents.RegistryIDs() {
		label := string(id)
		if agents.Registry[id].Kind == agents.StrategyNone {
			label = string(id) + " (no multi-root)"
		}
		options = append(options, huh.NewOption(label, string(id)))
	}
	options = append(options, huh.NewOption("custom (configure addDirArgs)", agentSelectCustom))

	initial := agentSelectUnset
	if in.CurrentAgent != "" {
		initial = string(in.CurrentAgent)
	} else if len(in.CurrentAddDirArgs) > 0 {
		initial = agentSelectCustom
	}

	choice := initial
	if err := huh.NewSelect[string]().
		Title(in.Title).
		Options(options...).
		Value(&choice).
		Run(); err != nil {
		if errors.Is(err, huh.ErrUserAborted) {
			return AgentChoice{Cancelled: true}, nil
		}
		return AgentChoice{Cancelled: true}, err
	}

	switch choice {
	case agentSelectUnset:
		return AgentChoice{}, nil
	case agentSelectCustom:
		template := "--add-dir {path}"
		if len(in.CurrentAddDirArgs) > 0 {
			template = strings.Join(in.CurrentAddDirArgs, " ")
		}
		if err := huh.NewInput().
			Title("addDirArgs template (space-separated, use {path} for sibling)").
			Placeholder("--add-dir {path}").
			Value(&template).
			Validate(func(s string) error {
				if !strings.Contains(s, "{path}") {
					return errors.New("must contain {path}")
				}
				return nil
			}).
			Run(); err != nil {
			if errors.Is(err, huh.ErrUserAborted) {
				return AgentChoice{Cancelled: true}, nil
			}
			return AgentChoice{Cancelled: true}, err
		}
		return AgentChoice{AddDirArgs: ParseLaunchCommand(template)}, nil
	default:
		return AgentChoice{Agent: agents.AgentID(choice)}, nil
	}
}
